source/Depreciation.cpp

```diff
@@ -63,6 +63,41 @@ double Depreciation::Full()
 
 
 
+// Calculate the value fraction for an item of the given age.
+double Depreciation::Depreciate(int age)
+{
+	if(age <= GracePeriod())
+		return 1.;
+
+	if(age >= MaxAge())
+		return Min();
+
+	double daily = pow(Daily(), age - GracePeriod());
+	double linear = static_cast<double>(MaxAge() - age) / (MaxAge() - GracePeriod());
+	return Min() + (1. - Min()) * daily * linear;
+}
+
+
+
+// Calculate how old an item needs to be for a target amount of depreciation, if possible.
+int Depreciation::AgeForDepreciation(double depreciation)
+{
+	if(depreciation <= Min())
+		return MaxAge();
+
+	if(depreciation >= 1)
+		return 0;
+
+	// Unfortunately, solving the depreciation function for age gives a horrible formula involving
+	// the Lambert W function. So instead, just increase age by 1 until the depreciation is right.
```

Comment:
Doesn't the comment say that's hard to do?

Reply:
Right right but still precomputing the vector sounds better to me
Especially since it'd be stored in decreasing order, surely there's a way to find stuff very quickly?

Reply:
why is this resolved?

Reply:
> we could prepare the values for the whole vector once and then just find it instantly (I think after one year the max is reached?)

The problem is that there are a lot of possible values.
The normal max is reached after one year, but you can make the price go lower past the max in the code if you want. Currently that is possible with this PR, and so there isn't really a maximum.

So I don't think it would be better to switch, because it wouldn't be faster and is there really another upside?

Reply:
> > The normal max is reached after one year, but you can make the price go lower past the max in the code if you want.
> 
> Why? We don't want to be selling random stock for too cheap lest it becomes overpowered.

Isn't it the job of the content creators to make sure they don't make it too cheap?
I imagine there are some plugins which would love to make Archons available from day 1 of a new save.
At the moment, the player wouldn't be able to buy them, but if the plugin author can set the price much lower...

> Even the 25% cost that is the current limit, you can get some amazing (perhaps too amazing) deals. I don't see why we would want them to be even cheaper. (And if they do get cheaper, can you sell them for under 25% of the price? Some other code might cap the value back to that preset limit.)

I don't think there's anything that will lower the amount of depreciation.

> Precalculating the costs for this range shouldn't be too difficult. It might even be doable in compile-time, assuming the depreciation gamerules don't mess that up.

I think they might.

> (At the very least, we could make use of the fact that depreciation only ever decreases the value, so we could search through it in logarithmic time - which, in this case, is a substantial difference, as this is an expensive operation.)

If it's expensive, then should it really happen?